### ISSUE TITLE
Data manipulation is way slower compared to DB call...

### ISSUE THREAD

**Issue by JulienPeloton** [astrolabsoftware/fink-science-portal#100]
... and this needs to be improved!

####  example
query: date search 2019-11-03 02:40:00 for 60 minutes window

# objects: 17625

call:  3.015425205230713 [seconds]
pandas:  12.757916927337646
classification:  0.029796600341796875
color (fake see #99):  0.034826040267944336
timestamp:  5.689717054367065
total time:  24.903888940811157

```python
# api.py - this takes 12 seconds!
pdfs = pd.DataFrame.from_dict(hbase_output, orient='index')
```

**Comment by karpov-sv** [comment]
Yes, I also encountered this problem while profiling the portal - just plain conversion of Java `TreeMap` to Pandas takes significantly longer than HBase search + transfer.
It is not specific to Pandas - most of the time is spent in just iterating over Java objects (example timing is for ~1200 rows):
```
In [8]: %time pdf=pd.DataFrame.from_dict(results, orient='index')
CPU times: user 1.37 s, sys: 12.7 ms, total: 1.38 s
Wall time: 1.38 s

In [9]: %time pdf=pd.DataFrame.from_dict(results.values())
CPU times: user 1.36 s, sys: 0 ns, total: 1.36 s
Wall time: 1.36 s

In [10]: %time res = [dict(results[_]) for _ in results]
CPU times: user 897 ms, sys: 2.12 ms, total: 899 ms
Wall time: 897 ms
```

Maybe some other, not so dynamic, data format may be used to represent the results in Java, so that it may be converted to Python data type in bulk and much faster?

**Comment by JulienPeloton** [comment]
Actually, without thinking about changing the underlying data structure on the Java side, we can slightly modify the current code to be limited by the time to iterate over the Java objects, without the Pandas extra overhead (on ~1000 rows):

```python
In [3]: def convert_before(data):
   ...:     converted = {i: dict(j) for i, j in results.items()}
   ...:     pdf = pd.DataFrame.from_dict(converted, orient='index')
   ...:     return pdf

In [6]: %timeit pdf=pd.DataFrame.from_dict(results, orient='index')
1.33 s ± 19.8 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [7]: %timeit pdf = convert_before(results)
911 ms ± 13.1 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

25% better performance for free :-)

**Comment by JulienPeloton** [comment]
> Maybe some other, not so dynamic, data format may be used to represent the results in Java, so that it may be converted to Python data type in bulk and much faster?

I agree this would be a better long-term solution. I'm not enough knowledgeable on Java for this. Let's see if @hrivnac has an idea on this.

**Comment by karpov-sv** [comment]
> 25% better performance for free :-)

Yes, that helps a bit. Next point is the color computation -  it takes approximately the same time as the conversion, so is also worth optimizing...

**Comment by karpov-sv** [comment]
The optimization from above is included in PR #557

**Comment by karpov-sv** [comment]
So it seems JPype is indeed slow at iterative access to Java objects, so the idea is to extract some bulk thing from it atomically that may be then parsed back in Python.
```python
In [1]: %timeit pdf=pd.DataFrame.from_dict({i: dict(j) for i, j in results.items()}, orient='index')
9.76 s ± 59.8 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [2]: from org.json import JSONObject

In [3]: import json

In [4]: %timeit pd.DataFrame.from_dict(json.loads(JSONObject(results).toString()), orient='index')
1.79 s ± 73.9 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```
Probably that's the best we may do for now (except you know how to use instead some better and faster type-preserving intermediate format - Parquet?..).

Implemented in #559

**Comment by JulienPeloton** [comment]
Impressive speed-up!
